lib/sprinkler/devices/watering_policy.ex

```diff
@@ -0,0 +1,86 @@
+defmodule Sprinkler.Devices.WateringPolicy do
+  def should_irrigate(moisture_input, temperature_input, humidity_input) do
+    unless soil_moisture_wet?(moisture_input) do
+      when_soil_moisture_not_wet(moisture_input, temperature_input, humidity_input)
+    end
+  end
+
+  defp when_soil_moisture_not_wet(moisture_input, temperature_input, humidity_input) do
+    if soil_moisture_medium?(moisture_input) do
+      when_soil_moisture_medium(temperature_input, humidity_input)
+    else
+      when_soil_moisture_not_medium(temperature_input, humidity_input)
+    end
+  end
+
+  defp when_soil_moisture_medium(temperature_input, humidity_input) do
+    if ambience_temperature_high?(temperature_input) do
+      irrigate('medium')
+    else
+      unless humidity_high?(humidity_input) do
+        irrigate('low')
+      end
+    end
+  end
```

Comment:
can't you pattern match this?

Reply:
mmm, I don't understand the question.. what do you mean?